broadcast/src/linked/signer/actor.rs

```diff
@@ -0,0 +1,1010 @@
+                num_items += 1;
+                let (_, _, _, msg) = msg.expect("unable to decode journal message");
+                let link = wire::Link::decode(msg).expect("journal message is unexpected format");
+                let height = link.chunk.as_ref().unwrap().height;
+                match tip {
+                    None => {
+                        tip = Some(link);
+                    }
+                    Some(ref t) => {
+                        if height > t.chunk.as_ref().unwrap().height {
+                            tip = Some(link);
+                        }
+                    }
+                }
+            }
+
+            // Set the tip
+            if let Some(link) = tip.take() {
+                let is_new = self.tip_man.put(&link);
+                assert!(is_new);
+            }
+
+            debug!(
+                sequencer = hex(sequencer),
+                n = num_items,
+                "Journal replay complete"
+            );
+        }
+
+        // Store journal
+        self.journals.insert(sequencer.clone(), journal);
+    }
+
+    /// Write a link to the appropriate journal.
+    ///
+    /// The journal must already be open and replayed.
+    async fn journal_append(&mut self, link: &wire::Link) {
+        let chunk = link.chunk.as_ref().unwrap();
+        let section = self.get_journal_section(chunk.height);
+        self.journals
+            .get_mut(&chunk.sequencer)
+            .expect("journal does not exist")
+            .append(section, link.encode_to_vec().into())
+            .await
+            .expect("unable to append to journal");
+    }
+
+    /// Syncs (ensures all data is written to disk) and prunes the journal for the given sequencer and height.
+    async fn journal_sync(&mut self, sequencer: &PublicKey, height: u64) {
+        let section = self.get_journal_section(height);
+
+        // Get journal
+        let journal = self
+            .journals
+            .get_mut(sequencer)
+            .expect("journal does not exist");
+
+        // Sync journal
+        journal.sync(section).await.expect("unable to sync journal");
+
+        // Prune journal, ignoring errors
+        let _ = journal.prune(section).await;
+    }
+
+    ////////////////////////////////////////
+    // Epoch
+    ////////////////////////////////////////
+
+    /// Updates the epoch to the value of the coordinator, and sets the refresh epoch deadline.
+    fn refresh_epoch(&mut self) {
+        // Set the refresh epoch deadline
+        self.refresh_epoch_deadline = Some(self.runtime.current() + self.refresh_epoch_timeout);
+
+        // Ensure epoch is not before the current epoch
+        let epoch = self.coordinator.index();
+        if epoch < self.epoch {
+            panic!("epoch must be greater than or equal to the current epoch");
+        }
+
+        // Update the epoch
+        self.epoch = epoch;
+    }
+}
+
+/// Errors that can occur when running the actor.
```

Comment:
nit: usually, I put these in `mod.rs`

Reply:
I would if they were used externally to this file, but since they are only ingested within the actor itself, I just moved it into that file